[6x only]Introduce GUC `gp_enable_mdqa_shared_scan` for queries with multi-DQA

QingMa-gp opened the pull request:
Introduce GUC `gp_enable_mdqa_shared_scan ` for queries with multi-DQA

In some scenarios, using a shared scan during three-stage aggregation
can lead to significant disk I/O and generate large spill files. To
address this issue, a new GUC, `gp_enable_mdqa_shared_scan`, has been
introduced. This configuration option controls whether to use shared
scan in three-stage aggregations.

By default, `gp_enable_mdqa_shared_scan` is set to `true`, enabling
shared scans during three-stage aggregation. This default setting aims
to optimize performance in general use cases. However, for specific
scenarios where disk I/O and spill file size are concerns, users can now
disable shared scans by setting this GUC to `false`.

This change allows for more flexible performance tuning.

This change is 6x only. since GPDB7 will not generate similar query plans for multi DQA.
## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

kainwen reviewed (commented):
After some thoughts, I think:  this GUC set true, it also use cost model that might not force.

So can we introduce a `disallow_share_scan` GUC? Default false, if true, then just disallow.

Lets wait for more people's comments. Thanks a lot.

ashwinstar commented:
> After some thoughts, I think: this GUC set true, it also use cost model that might not force.
> So can we introduce a disallow_share_scan GUC? Default false, if true, then just disallow.

We should make it consistent with other similar purpose GUCs like `enable_indexscan`. The GUC says "Enables the planner's use of index-scan plans." which means it just enables planner to choose such plans, still planner depending on situation when this GUC is on may not pick indexscan. Setting it to FALSE for sure disables its usage. Hence, we should do similar here.

We should rename the GUC as don't need `gp` in the name. Maybe call it `enable_sharedscan_three_stage_agg`

ashwinstar commented on src/backend/utils/misc/guc_gp.c:
```diff
@@ -2075,6 +2076,16 @@ struct config_bool ConfigureNamesBool_gp[] =
 		true, NULL, NULL
 	},
 
+		{
```
Something we and upstream doesn't do good job currently, is providing guidance around usage of the GUC. We definitely need to improve on that. So would be good if some description text can be added to "Enable shared scan in three stage agg." to clarify in what situations (large spillfiles I guess in this situation) should someone decide to disable this functionality. What benefit it would provide, what could the downside be (using more memory or what?) and such aspects covered.

gpopt replied:
+1

gpopt commented on src/backend/utils/misc/guc_gp.c:
```diff
@@ -2075,6 +2076,16 @@ struct config_bool ConfigureNamesBool_gp[] =
 	},
 
+		{
+		{"gp_three_stage_agg_use_share", PGC_USERSET, QUERY_TUNING_METHOD,
```
The GUC name and the explanation is still confusing. Users hardly know what is three stage agg, it is actually referring to the query with multiple distinct aggregates, right?
If I understand correctly, it is better to name it `gp_enable_mdqa_shared_scan` or `gp_multi_distinct_agg_share_scan`, etc.
Also refine the explanation for the GUC.

QingMa-gp commented:
Hi, @ashwinstar @gpopt . Thanks for the suggestion, I have renamed the GUC to `gp_enable_mdqa_shared_scan` and updated the explanation, please help to review again.

kainwen commented on src/test/regress/expected/gp_aggregates.out:
```diff
@@ -648,3 +648,79 @@ GROUP BY bb.v;
 reset optimizer;
 reset enable_groupagg;
+-- test multi DQA with guc gp_enable_mdqa_shared_scan
+set optimizer = off; -- the case is planner only, so disable orca
```
Even this is planner only case, we can stil let keep optimizer unchanged. No harm.

kainwen replied:
Comment clear is enough.

kainwen replied:
Remind.

QingMa-gp replied:
We don't have `gp_aggregates_optimizer.out` for this test, so set the `optimizer` to off here as the previous testcase did.

QingMa-gp commented:
https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-three_stage_agg_use_share_6x-rocky8 The failed test is unrelated to this modifications.

ashwinstar commented on src/backend/utils/misc/guc_gp.c:
```diff
@@ -2075,6 +2076,16 @@ struct config_bool ConfigureNamesBool_gp[] =
 
+		{
+		{"gp_enable_mdqa_shared_scan", PGC_USERSET, QUERY_TUNING_METHOD,
+			gettext_noop("Planner Only. Enable shared scan for queries with multiple distinct aggregations, set it to \"false\" to avoid large spill file."),
```
"set it to \"false\" to avoid large spill file." is always desirable - so should also explain why someone should have this GUC set to TRUE at all which is default?

QingMa-gp replied:
I have updated the description accordingly.
